pkg/protocol/block_dispatcher.go

```diff
@@ -205,21 +214,130 @@ func (b *BlockDispatcher) processWarpSyncResponse(commitmentID iotago.Commitment
 		return ierrors.Errorf("failed to get target engine for %s", commitmentID)
 	}
 
+	// Make sure that already evicted commitments are not processed. This might happen if there's a lot of slots to process ]
+	// and old responses are still in the task queue.
+	if loadedCommitment, err := targetEngine.Storage.Commitments().Load(commitmentID.Slot()); err == nil && loadedCommitment.ID() == commitmentID {
+		return nil
+	}
+
 	acceptedBlocks := ads.NewSet[iotago.BlockID](mapdb.NewMapDB(), iotago.BlockID.Bytes, iotago.SlotIdentifierFromBytes)
 	for _, blockID := range blockIDs {
-		_ = acceptedBlocks.Add(blockID) // a mapdb can newer return an error
+		_ = acceptedBlocks.Add(blockID) // a mapdb can never return an error
+	}
+
+	if !iotago.VerifyProof(tangleMerkleProof, iotago.Identifier(acceptedBlocks.Root()), chainCommitment.Commitment().RootsID()) {
+		return ierrors.Errorf("failed to verify tangle merkle proof for %s", commitmentID)
+	}
+
+	acceptedTransactionIDs := ads.NewSet[iotago.TransactionID](mapdb.NewMapDB(), iotago.TransactionID.Bytes, iotago.TransactionIDFromBytes)
+	for _, transactionID := range transactionIDs {
+		_ = acceptedTransactionIDs.Add(transactionID) // a mapdb can never return an error
 	}
 
-	if !iotago.VerifyProof(merkleProof, iotago.Identifier(acceptedBlocks.Root()), chainCommitment.Commitment().RootsID()) {
-		return ierrors.Errorf("failed to verify merkle proof for %s", commitmentID)
+	if !iotago.VerifyProof(mutationMerkleProof, iotago.Identifier(acceptedTransactionIDs.Root()), chainCommitment.Commitment().RootsID()) {
+		return ierrors.Errorf("failed to verify mutation merkle proof for %s", commitmentID)
 	}
 
 	b.pendingWarpSyncRequests.StopTicker(commitmentID)
 
 	b.processedWarpSyncRequests.Add(commitmentID)
```

Comment:
Should these two operations be atomic to prevent races in processing & requesting?

Reply:
Currently there can't be any races because there's only 1 worker in `warpSyncWorkers` processing the warpsync responses/requests. And making this atomic would also not help to prevent the following scenario:
You receive multiple warp sync responses for the same slot (either from the same neighbor or different ones, doesn't really matter) and then they will be queued already. Some of these might be old and will be already evicted from `processedWarpSyncRequests` map and thus we don't even remember anymore that we processed said commitment. That's why I needed to add following to prevent processing an old commitment again:
<img width="1155" alt="image" src="https://github.com/iotaledger/iota-core/assets/4181434/e3cee97d-45dd-41fe-a109-479784ef95d9">